### PR TITLE
[WIP] Adds Controller and wraps functionality into widget

### PR DESCRIPTION
**Pull request by Nolence**
This PR is largely a rewrite but keeps most of the line drawing, constraint and position finding in place. An issue where the tooltip would not follow the child in a listview upon scrolling has also been solved. I've also favored largely favored functional programming and made several changes to functions and fields. Most comments have been removed for legibility but will soon be added back. I just wanted your take on the implementation as well as a few features I plan to add.

How involved would it be in `getPositionForChild` to switch the preferred direction if the child does not have sufficient space to fit in the allotted constraints. The native default `Tooltip` from Flutter does this using [`positionDependentBox`](https://api.flutter.dev/flutter/painting/positionDependentBox.html).

Why does having a borderWidth of 0.0 still draw a border? Is this some flutter `draw` quirk?

Sorry to completely butcher your code renaming things left and write. It was supposed to just be a custom implementation until I decided to try a PR

### CODE REVIEW

**Comment by Nolence** [comment]
There might be some things I forgot to add back in (such as close button). I'll make sure to add those in later

File: lib/super_tooltip.dart [hunk 1]
```diff
@@ -1,908 +1,1084 @@
...
+    this.onLongPress,
+    this.onShow,
+    this.onHide,
+    this.snapsFarAwayVertically = false,
+    this.snapsFarAwayHorizontally = false,
     this.top,
     this.right,
     this.bottom,
     this.left,
-    this.minimumOutSidePadding = 20.0,
```
Comment (escamoteur) [inline lib/super_tooltip.dart]: Why did you remove them?

File: lib/super_tooltip.dart [hunk 2]
```diff
@@ -1,908 +1,1084 @@
...
-    @required this.content, // The contents of the tooltip.
-    @required this.popupDirection,
-    this.onClose,
-    this.minWidth,
-    this.minHeight,
-    this.maxWidth,
-    this.maxHeight,
+  Future<void> hideTooltip() {
+    event = _Event.hide;
+    _completer = Completer();
```
[the remark this answers is not in the document]
Comment (Nolence) [reply lib/super_tooltip.dart]: I call the completer when i call either `_showTooltip` or `_hideTooltip`.

```dart
_showTooltip() async {
  widget.onShow?.call();

  if (_entry != null) return; // Already visible.

  _createOverlayEntries();

  await _animationController
      .forward()
      .whenComplete(_superTooltipController.complete);
}
```

After the animation controller has finished animating the widget either in or out, calling `_superTooltipController.complete` will toggle the `isVisible` field in the controller. The idea was partially taken from [transformer_page_view](https://github.com/best-flutter/transformer_page_view/blob/master/lib/index_controller.dart) plugin.

Let me know if it's necessary or not. I haven't found a use case but thought it might be expected in the controller.

File: lib/super_tooltip.dart (hunk 1 above)
Comment (Nolence) [reply lib/super_tooltip.dart]: The word padding didn't sound appropriate to me, as it implied the spacing inside of the widget as opposed to a margin, which is spacing outside the widget. Do you agree? As for the `minimumOutside` prior to the word, sorry, I just forgot to add that. Will add it back in next commit

File: lib/super_tooltip.dart [hunk 3]
```diff
@@ -1,908 +1,1084 @@
...
-  /// Removes the Tooltip from the overlay
-  void close() {
-    if (onClose != null) {
-      onClose();
-    }
+    this.showCloseButton = ShowCloseButton.none,
+    // FIXME: Why does a 0.0 borderwidth still show a border?
+    this.borderWidth = 0.0,
+    this.borderRadius = 10.0,
+  }) : super(key: key);
```
[the remark this answers is not in the document]
Comment (Nolence) [reply lib/super_tooltip.dart]: No worries! I'm scared to touch the path drawing stuff but I'll have a look this weekend

[the remark this answers is not in the document]
Comment (escamoteur) [reply lib/super_tooltip.dart]: I can imagine. It was a lot of work getting everthing right

File: lib/super_tooltip.dart (hunk 1 above)
Comment (escamoteur) [reply lib/super_tooltip.dart]: yeah you are right margin is a better term although technically its a padding to the wrapping container.

File: lib/super_tooltip.dart [hunk 4]
```diff
@@ -1,908 +1,1084 @@
...
-    Overlay.of(targetContext).insertAll([_backGroundOverlay, _ballonOverlay]);
-    isOpen = true;
+  @override
+  void initState() {
+    _animationController = AnimationController(
+      duration: widget.fadeInDuration,
+      reverseDuration: widget.fadeOutDuration,
+      vsync: this,
+    )..addStatusListener(_handleStatusChanged);
+    _superTooltipController = widget.controller ?? SuperTooltipController();
```
[the remark this answers is not in the document]
Comment (Nolence) [reply lib/super_tooltip.dart]: If the user doesn't provide a controller, I'm assuming they don't want access to it. They still have access to `onShow` and `onHide`, etc. Never tried to implement a controller so I could be all wrong

[the remark this answers is not in the document]
Comment (Nolence) [reply lib/super_tooltip.dart]: With or without a controller, clicking on the child will open the tooltip and add a slightly transparent barrier behind it. Clicking on the barrier will close the tooltip. I also plan to add support for mouse clicking/hovering as well as adding a automatic show/hide timers similar to the [original tooltip](https://github.com/flutter/flutter/blob/57160b938cd4e1762d33bbf71914d6e86c95fe18/packages/flutter/lib/src/material/tooltip.dart#L193).

**Comment by Nolence** [comment]
The static function is a great idea. I'll have a look into that later. Gotta get back to ft work currently

File: lib/super_tooltip.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (escamoteur) [reply lib/super_tooltip.dart]: Ah, OK now I recall to have seen a GestureDetector. That's different to the original one. Can this be diesabled by the controller?

**Comment by escamoteur** [comment]
Hi,

how should we proceed now?

**Comment by Nolence** [comment]
Sorry @escamoteur, got sidetracked by work and I've also been trying to finish a personal project by the end of this month. I'll dedicate a day next week to updating this PR with my latest work though -- as I think have made one improvement in terms of drawing the tooltip shape. I'm not sure on how robust it is but it significantly less lines. Tell me what you think. THe gist is [here](https://gist.github.com/Nolence/d07a1e04877666fb64604c2d710f2906)

**Comment by Nolence** [comment]
It's a regression in terms of functionality as it only does top and bottom, but left and right shouldn't be too hard to add.

**Comment by Nolence** [comment]
Sorry, I don't see myself finishing this any time soon 😞. My custom implementation diverged a lot more from what I needed from this library and they hardly resemble one another now. This new job has me on a time crunch  but when I'm finished with that I'll hopefully reopen
